hoot-core/src/main/cpp/hoot/core/io/OsmJsonReader.cpp

```diff
@@ -0,0 +1,410 @@
+ */
+
+#include "OsmJsonReader.h"
+
+// hoot
+#include <hoot/core/Hoot.h>
+#include <hoot/core/util/HootException.h>
+#include <hoot/core/util/Log.h>
+#include <hoot/core/Factory.h>
+
+// Boost
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/foreach.hpp>
+namespace pt = boost::property_tree;
+
+// Qt
+#include <QTextCodec>
+#include <QEventLoop>
+#include <QTextStream>
+#include <QtNetwork/QNetworkReply>
+#include <QtNetwork/QNetworkRequest>
+#include <QtNetwork/QNetworkAccessManager>
+
+// Standard
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <unistd.h>
+using namespace std;
+
+namespace hoot
+{
+
+HOOT_FACTORY_REGISTER(OsmMapReader, OsmJsonReader)
+
+// Default constructor
+OsmJsonReader::OsmJsonReader():
+  _defaultStatus(Status::Invalid),
+  _useDataSourceIds(true),
+  _defaultCircErr(15.0),
+  _propTree(),
+  _version(""),
+  _generator(""),
+  _timestamp_base(""),
+  _copyright(""),
+  _url(""),
+  _isFile(false),
+  _isWeb(false)
+{
+  // Do nothing special
+}
+
+OsmJsonReader::~OsmJsonReader()
+{
+  close();
+}
+
+bool OsmJsonReader::isSupported(QString url)
+{
+  QUrl myUrl(url);
+
+  // Is it a file?
+  if (myUrl.isLocalFile())
+  {
+    QString filename = myUrl.toLocalFile();
+
+    if (QFile::exists(filename))
+      return true;
+  }
+
+  // Is it a web address?
+  if ("http" == myUrl.scheme() || "https" == myUrl.scheme())
+  {
+    return true;
+  }
+
+  // Default to not supported
+  return false;
+}
+
+/**
+ * Opens the specified URL for reading.
+ */
+void OsmJsonReader::open(QString url)
+{
+  try
+  {
+    // Bail out if unsupported
+    if (!isSupported(url))
+      return;
+
+    // Handle files or URLs
+    _url = QUrl(url);
+    if (_url.isLocalFile())
+    {
+      _isFile = true;
+      _file.setFileName(_url.toLocalFile());
+      _file.open(QFile::ReadOnly | QFile::Text);
+    }
+    else
+    {
+      _isWeb = true;
+    }
+  }
+  catch (std::exception ex)
+  {
+    ostringstream oss;
+    oss << "Exception opening URL (" << url << "): " << ex.what();
+    throw HootException(QString::fromStdString(oss.str()));
+  }
+}
+
+void OsmJsonReader::close()
+{
+  if (_isFile)
+    _file.close();
+}
+
+/**
+ * Reads the specified map. When this method is complete
+ * the input will likely be closed.
+ */
+void OsmJsonReader::read(shared_ptr<OsmMap> map)
+{
+  if (_isFile)
+  {
+    QTextStream instream(&_file);
+    QString jsonStr = instream.readAll();
+    _loadJSON(jsonStr);
+    _parseOverpassJson(map);
+  }
+  else
+  {
+    boost::shared_ptr<QNetworkAccessManager> pNAM(new QNetworkAccessManager());
+    QNetworkRequest request(_url);
+    boost::shared_ptr<QNetworkReply> pReply(pNAM->get(request));
+
+    // Wait for finished signal from reply object
+    QEventLoop loop;
+    QObject::connect(pReply.get(), SIGNAL(finished()), &loop, SLOT(quit()));
+    loop.exec();
+
+    QByteArray data = pReply->readAll();
```

Comment:
Feeling particularly motivated? I love it!

You'll want to check the error status here and throw an exception. In general Qt doesn't throw exceptions. (`QNetworkReply::error()`)